File: six/modules/python/cphd03/tests/test_vbm.py
#!/usr/bin/env python

#
# =========================================================================
# This file is part of cphd03-python
# =========================================================================
#
# (C) Copyright 2004 - 2015, MDA Information Systems LLC
#
# cphd03-python is free software; you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this program; If not,
# see <http://www.gnu.org/licenses/>.
#

# In general, if functionality in CPHD is borrowed from six.sicd,
# refer to six.sicd's test script for more verification

# The numbers used here don't correspond to anything at all

from pysix import cphd03
import pysix.six_base as six
from pysix import six_sicd
from pysix import scene
from coda import coda_types
from coda import math_poly
from coda import math_linear
from coda import coda_io
import numpy

def notEqualVector3(lhs, rhs):
    return lhs[0] != rhs[0] or lhs[1] != rhs[1] or lhs[2] != rhs[2]

if __name__ == '__main__':

    vecParam = cphd03.VectorParameters()

    domainType = cphd03.DomainType()
    domainType.value = cphd03.DomainType.TOA

    numVec = coda_types.VectorSizeT()
    numVec.push_back(3)
    numVec.push_back(3)
    numVec.push_back(3)
    vbm = cphd03.VBM(3,
            numVec,
            True,
            False,
            True,
            domainType 
            )

    if vbm.haveTropoSRP():
        print 'VBM TropoSRP selection not working'

    if not vbm.haveAmpSF():
        print 'VBM AmpSF selection not workig'

    if not vbm.haveSRPTime():
        print 'VBM SRPTime selection not working'

    assignNum = 99
    vbm.setTxTime(assignNum,1,1)
    if vbm.getTxTime(1,1) != assignNum:
        print 'VBM txtime assign failed'


    assignVec = math_linear.Vector3()
    assignVec[0] = 1
    assignVec[1] = 2
    assignVec[2] = 3
    vbm.setTxPos(assignVec,2,2)
    if notEqualVector3(assignVec, vbm.getTxPos(2,2)):
        print 'VBM txpos assign failed'

    vbm.setRcvTime(assignNum, 0, 1)
    if vbm.getRcvTime(0,1) != assignNum:
        print 'VBM rcvtime assign failed'

    vbm.setRcvPos(assignVec,0,0)
    if notEqualVector3(vbm.getRcvPos(0,0), assignVec):
        print 'VBM rcvpos assign failed'

    vbm.setSRPTime(assignNum, 1, 0)
    if vbm.getSRPTime(1,0) != assignNum:
        print 'VBM srptime assign failed'

    vbm.setSRPPos(assignVec, 2, 0)
    if notEqualVector3(vbm.getSRPPos(2,0), assignVec):
        print 'VBM srppos assign failed'

    vbm.setDeltaTOA0(assignNum, 1,2)
    if vbm.getDeltaTOA0(1,2) != assignNum:
        print 'VBM deltaTOA0 assign failed'

    vbm.setTOASS(assignNum, 1,1)
    if vbm.getTOASS(1,1) != assignNum:
        print 'VBM TOASS assign failed'

    vbm.updateVectorParameters(vecParam)

    # If you want to be sure things are getting assigned,
    # just uncomment all these

    # These show that our vecParam was updated using our optional
    # field selection (tropoSRP size will be undefined, tropoSRP offset
    # will be negative 1)
    #print vecParam.txTime
    #print vecParam.txPos
    #print vecParam.rcvTime
    #print vecParam.rcvPos
    #print vecParam.srpTime
    #print vecParam.srpPos
    #print vecParam.tropoSRP
    #print vecParam.ampSF
    #print vecParam.toaParameters.deltaTOA0
    #print vecParam.toaParameters.toaSS

    #print vecParam.txTimeOffset()
    #print vecParam.txPosOffset()
    #print vecParam.rcvTimeOffset()
    #print vecParam.rcvPosOffset()
    #print vecParam.srpTimeOffset()
    #print vecParam.srpPosOffset()
    #print vecParam.tropoSRPOffset()
    #print vecParam.ampSFOffset()
    #print vecParam.deltaTOA0Offset()
    #print vecParam.toaSSOffset()

    #
    # Now we'll do a round trip test:
    #
    # Print everything out, load it up with VBM's
    # load() function and make sure it matches
    #

    testFileName = 'test.vbm'

    ch0 = vbm.toBuffer(0)
    ch1 = vbm.toBuffer(1)
    ch2 = vbm.toBuffer(2)

    # byteswap and write
    testFile = open(testFileName,'wb')
    testFile.write(ch0.byteswap())
    testFile.write(ch1.byteswap())
    testFile.write(ch2.byteswap())
    testFile.close()

    newVbm = cphd03.VBM(3,numVec, True, False, True, domainType)
    fis = coda_io.FileInputStream(testFileName)

    # sum of size of each channel
    vbmSize = vbm.getVBMsize(0) + vbm.getVBMsize(1) + vbm.getVBMsize(2)

    retv = newVbm.load(fis, 0, vbmSize, 1)
    if retv != vbmSize:
        print 'VBM read failed'

    if fis.isOpen():
        fis.close()

    nCh0 = newVbm.toBuffer(0)
    nCh1 = newVbm.toBuffer(1)
    nCh2 = newVbm.toBuffer(2)

    if not numpy.array_equal(ch0, nCh0):
        print 'Channel 0 load failed'

    if not numpy.array_equal(ch1, nCh1):
        print 'Channel 1 load failed'

    if not numpy.array_equal(ch2, nCh2):
        print 'Channel 2 load failed'

File: six/modules/c++/cphd03/unittests/test_vbm.cpp
/* =========================================================================
 * This file is part of cphd03-c++
 * =========================================================================
 *
 * (C) Copyright 2004 - 2014, MDA Information Systems LLC
 *
 * cphd03-c++ is free software; you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this program; If not,
 * see <http://www.gnu.org/licenses/>.
 *
 */

#include <string>
#include <tuple>

#include <cphd03/VBM.h>
#include <cphd/TestDataGenerator.h>
#include "TestCase.h"

static constexpr size_t NUM_CHANNELS = 3;
static constexpr size_t NUM_VECTORS = 7;

static void call_srand()
{
    static const auto f = []() { constexpr auto seed = 174; ::srand(seed); return nullptr; };
    static const auto result = f();
    std::ignore = result;
}

void testVectorParameters(const std::string& testName_,
                          size_t channel,
                          size_t vector,
                          cphd03::VBM& vbm)
{
    const auto testName = testName_;

    const double txTime = cphd::getRandom();
    vbm.setTxTime(txTime, channel, vector);

    const cphd::Vector3 txPos = cphd::getRandomVector3();
    vbm.setTxPos(txPos, channel, vector);

    const double rcvTime = cphd::getRandom();
    vbm.setRcvTime(rcvTime, channel, vector);

    const cphd::Vector3 rcvPos = cphd::getRandomVector3();
    vbm.setRcvPos(rcvPos, channel, vector);

    const double srpTime = cphd::getRandom();
    vbm.setSRPTime(srpTime, channel, vector);

    const cphd::Vector3 srpPos = cphd::getRandomVector3();
    vbm.setSRPPos(srpPos, channel, vector);

    const double tropoSrp = cphd::getRandom();
    vbm.setTropoSRP(tropoSrp, channel, vector);

    const double ampSF = cphd::getRandom();
    vbm.setAmpSF(ampSF, channel, vector);

    TEST_ASSERT_EQ(txTime, vbm.getTxTime(channel, vector));
    TEST_ASSERT_EQ(txPos, vbm.getTxPos(channel, vector));
    TEST_ASSERT_EQ(rcvTime, vbm.getRcvTime(channel, vector));
    TEST_ASSERT_EQ(rcvPos, vbm.getRcvPos(channel, vector));
    TEST_ASSERT_EQ(srpTime, vbm.getSRPTime(channel, vector));
    TEST_ASSERT_EQ(srpPos, vbm.getSRPPos(channel, vector));
    TEST_ASSERT_EQ(tropoSrp, vbm.getTropoSRP(channel, vector));
    TEST_ASSERT_EQ(ampSF, vbm.getAmpSF(channel, vector));
}

TEST_CASE(testVbmFx)
{
    call_srand();

    cphd03::VBM vbm(NUM_CHANNELS,
                  std::vector<size_t>(NUM_CHANNELS, NUM_VECTORS),
                  true,
                  true,
                  true,
                  cphd::DomainType::FX);

    for (size_t channel = 0; channel < NUM_CHANNELS; ++channel)
    {
        for (size_t vector = 0; vector < NUM_VECTORS; ++vector)
        {
            testVectorParameters(testName, channel, vector, vbm);

            const double fx0 = cphd::getRandom();
            vbm.setFx0(fx0, channel, vector);

            const double fxSS = cphd::getRandom();
            vbm.setFxSS(fxSS, channel, vector);

            const double fx1 = cphd::getRandom();
            vbm.setFx1(fx1, channel, vector);

            const double fx2 = cphd::getRandom();
            vbm.setFx2(fx2, channel, vector);

            TEST_ASSERT_EQ(fx0, vbm.getFx0(channel, vector));
            TEST_ASSERT_EQ(fxSS, vbm.getFxSS(channel, vector));
            TEST_ASSERT_EQ(fx1, vbm.getFx1(channel, vector));
            TEST_ASSERT_EQ(fx2, vbm.getFx2(channel, vector));
        }
    }
}

TEST_CASE(testVbmToa)
{
    call_srand();

    cphd03::VBM vbm(NUM_CHANNELS,
                  std::vector<size_t>(NUM_CHANNELS, NUM_VECTORS),
                  true,
                  true,
                  true,
                  cphd::DomainType::TOA);

    for (size_t channel = 0; channel < NUM_CHANNELS; ++channel)
    {
        for (size_t vector = 0; vector < NUM_VECTORS; ++vector)
        {
            testVectorParameters("testVbmToa", channel, vector, vbm);

            const double deltaToa0 = cphd::getRandom();
            vbm.setDeltaTOA0(deltaToa0, channel, vector);

            const double toaSS = cphd::getRandom();
            vbm.setTOASS(toaSS, channel, vector);

            TEST_ASSERT_EQ(deltaToa0, vbm.getDeltaTOA0(channel, vector));
            TEST_ASSERT_EQ(toaSS, vbm.getTOASS(channel, vector));
        }
    }
}

TEST_CASE(testVbmThrow)
{
    call_srand();

    cphd03::VBM vbmFx(NUM_CHANNELS,
                    std::vector<size_t>(NUM_CHANNELS, NUM_VECTORS),
                    false,
                    false,
                    false,
                    cphd::DomainType::FX);

    cphd03::VBM vbmToa(NUM_CHANNELS,
                     std::vector<size_t>(NUM_CHANNELS, NUM_VECTORS),
                     false,
                     false,
                     false,
                     cphd::DomainType::TOA);

    for (size_t channel = 0; channel < NUM_CHANNELS; ++channel)
    {
        for (size_t vector = 0; vector < NUM_VECTORS; ++vector)
        {
            TEST_EXCEPTION(vbmFx.setSRPTime(0.0, channel, vector));
            TEST_EXCEPTION(vbmToa.setSRPTime(0.0, channel, vector));
            TEST_EXCEPTION(vbmFx.getSRPTime(channel, vector));
            TEST_EXCEPTION(vbmToa.getSRPTime(channel, vector));

            TEST_EXCEPTION(vbmFx.setTropoSRP(0.0, channel, vector));
            TEST_EXCEPTION(vbmToa.setTropoSRP(0.0, channel, vector));
            TEST_EXCEPTION(vbmFx.getTropoSRP(channel, vector));
            TEST_EXCEPTION(vbmToa.getTropoSRP(channel, vector));

            TEST_EXCEPTION(vbmFx.setAmpSF(0.0, channel, vector));
            TEST_EXCEPTION(vbmToa.setAmpSF(0.0, channel, vector));
            TEST_EXCEPTION(vbmFx.getAmpSF(channel, vector));
            TEST_EXCEPTION(vbmToa.getAmpSF(channel, vector));

            TEST_EXCEPTION(vbmFx.setDeltaTOA0(0.0, channel, vector));
            TEST_EXCEPTION(vbmFx.getDeltaTOA0(channel, vector));

            TEST_EXCEPTION(vbmFx.setTOASS(0.0, channel, vector));
            TEST_EXCEPTION(vbmFx.getTOASS(channel, vector));

            TEST_EXCEPTION(vbmToa.setFx0(0.0, channel, vector));
            TEST_EXCEPTION(vbmToa.getFx0(channel, vector));

            TEST_EXCEPTION(vbmToa.setFxSS(0.0, channel, vector));
            TEST_EXCEPTION(vbmToa.getFxSS(channel, vector));

            TEST_EXCEPTION(vbmToa.setFx1(0.0, channel, vector));
            TEST_EXCEPTION(vbmToa.getFx1(channel, vector));

            TEST_EXCEPTION(vbmToa.setFx2(0.0, channel, vector));
            TEST_EXCEPTION(vbmToa.getFx2(channel, vector));
        }
    }
}

TEST_CASE(testVbmCopy)
{
    call_srand();

    cphd03::VBM vbmFx(NUM_CHANNELS,
                    std::vector<size_t>(NUM_CHANNELS, NUM_VECTORS),
                    false,
                    false,
                    false,
                    cphd::DomainType::FX);

    cphd03::VBM vbmToa(NUM_CHANNELS,
                     std::vector<size_t>(NUM_CHANNELS, NUM_VECTORS),
                     false,
                     false,
                     false,
                     cphd::DomainType::TOA);

    cphd03::VBM vbmFxCopy(vbmFx);
    cphd03::VBM vbmToaCopy(vbmToa);

    TEST_ASSERT_EQ(vbmFx, vbmFxCopy);
    TEST_ASSERT_EQ(vbmToa, vbmToaCopy);
}

TEST_CASE(testDataConstructor)
{
    call_srand();

    std::vector<std::vector<double> > actualData(NUM_CHANNELS);
    const size_t numValuesNeeded = 18;
    for (size_t ii = 0; ii < NUM_CHANNELS; ++ii)
    {
        actualData[ii].resize(numValuesNeeded * NUM_VECTORS);
        for (size_t jj = 0; jj < NUM_VECTORS; ++jj)
        {
            for (size_t kk = 0; kk < numValuesNeeded; ++kk)
            {
                actualData[ii][(jj * numValuesNeeded) + kk] =
                        (ii * numValuesNeeded) + kk;
            }
        }
    }

    std::vector<const void*> data(NUM_CHANNELS);
    for (size_t ii = 0; ii < NUM_CHANNELS; ++ii)
    {
        data[ii] = actualData[ii].data();
    }

    cphd03::VBM vbm(NUM_CHANNELS,
                  std::vector<size_t>(NUM_CHANNELS, NUM_VECTORS),
                  true,
                  true,
                  true,
                  cphd::DomainType::FX,
                  data);

    for (size_t ii = 0; ii < NUM_CHANNELS; ++ii)
    {
        for (size_t jj = 0; jj < NUM_VECTORS; ++jj)
        {
            const size_t idx = jj * numValuesNeeded;
            TEST_ASSERT_ALMOST_EQ(actualData[ii][idx], vbm.getTxTime(ii, jj));
            const cphd::Vector3 txPos = vbm.getTxPos(ii, jj);
            TEST_ASSERT_ALMOST_EQ(actualData[ii][idx + 1], txPos[0]);
            TEST_ASSERT_ALMOST_EQ(actualData[ii][idx + 2], txPos[1]);
            TEST_ASSERT_ALMOST_EQ(actualData[ii][idx + 3], txPos[2]);

            TEST_ASSERT_ALMOST_EQ(actualData[ii][idx + 4], vbm.getRcvTime(ii, jj));
            const cphd::Vector3 rcvPos = vbm.getRcvPos(ii, jj);
            TEST_ASSERT_ALMOST_EQ(actualData[ii][idx + 5], rcvPos[0]);
            TEST_ASSERT_ALMOST_EQ(actualData[ii][idx + 6], rcvPos[1]);
            TEST_ASSERT_ALMOST_EQ(actualData[ii][idx + 7], rcvPos[2]);

            TEST_ASSERT_ALMOST_EQ(actualData[ii][idx + 8], vbm.getSRPTime(ii, jj));
            const cphd::Vector3 srpPos = vbm.getSRPPos(ii, jj);
            TEST_ASSERT_ALMOST_EQ(actualData[ii][idx + 9], srpPos[0]);
            TEST_ASSERT_ALMOST_EQ(actualData[ii][idx + 10], srpPos[1]);
            TEST_ASSERT_ALMOST_EQ(actualData[ii][idx + 11], srpPos[2]);

            TEST_ASSERT_ALMOST_EQ(actualData[ii][idx + 12], vbm.getTropoSRP(ii, jj));
            TEST_ASSERT_ALMOST_EQ(actualData[ii][idx + 13], vbm.getAmpSF(ii, jj));

            TEST_ASSERT_ALMOST_EQ(actualData[ii][idx + 14], vbm.getFx0(ii, jj));
            TEST_ASSERT_ALMOST_EQ(actualData[ii][idx + 15], vbm.getFxSS(ii, jj));
            TEST_ASSERT_ALMOST_EQ(actualData[ii][idx + 16], vbm.getFx1(ii, jj));
            TEST_ASSERT_ALMOST_EQ(actualData[ii][idx + 17], vbm.getFx2(ii, jj));
        }
    }
}

TEST_MAIN(
    TEST_CHECK(testVbmFx);
    TEST_CHECK(testVbmToa);
    TEST_CHECK(testVbmThrow);
    TEST_CHECK(testVbmCopy);
    TEST_CHECK(testDataConstructor);
    )

